File: python/samples/08_log_service.py
#!/usr/bin/env python3
"""Sample 08 — LogService: query with $top/$skip/$filter, pagination.

Demonstrates:
  - ctx.logs.list_services_async()
  - LogFilter with top, skip, severity, message_id (OData order enforced)
  - ctx.logs.get_entries_async() — single page
  - ctx.logs.iter_entries_async() — follow Members@odata.nextLink
  - ctx.logs.get_entry_async()
  - ctx.logs.clear_log_async()

Usage:
    python 08_log_service.py [--host HOST] [--port PORT]
"""

from __future__ import annotations

import argparse
import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import redfish_sdk
from redfish_sdk import AuthMode, ConnectionConfig, Credentials
from redfish_sdk.services.log_service import LogFilter
from redfish_sdk.errors import RedfishConnectionError


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Sample 08 — Log Service")
    p.add_argument("--host", default="127.0.0.1")
    p.add_argument("--port", type=int, default=8000)
    p.add_argument("--user", default="admin")
    p.add_argument("--password", default="password")
    p.add_argument("--no-tls-verify", action="store_true")
    p.add_argument("--max-entries", type=int, default=5, help="Max entries to display per log")
    return p.parse_args()


def _print_entry(entry: dict) -> None:
    eid      = entry.get("Id", "?")
    severity = entry.get("Severity", entry.get("MessageSeverity", "?"))
    msg      = entry.get("Message", "")[:72]
    created  = entry.get("Created", "")[:19]
    print(f"      [{eid:>10}] {severity:<12} {created}  {msg}")


async def main() -> None:
    args = parse_args()

    creds = Credentials(username=args.user, password=args.password)
    config = ConnectionConfig(verify_tls=not args.no_tls_verify)

    try:
        ctx = await redfish_sdk.connect_async(
            host=args.host,
            port=args.port,
            credentials=creds,
            auth_mode=AuthMode.SESSION,
            config=config,
        )
    except RedfishConnectionError as exc:
        print(f"[ERROR] {exc}")
        sys.exit(1)

    async with ctx:
        logs = ctx.logs

        # ── List log services ──────────────────────────────────────────────
        print("Listing available log services …")
        svc_resp = await logs.list_services_async()
        if not svc_resp.success:
            print(f"  ✗ HTTP {svc_resp.status_code}")
            return

        services = svc_resp.body.get("Members", [])
        print(f"  Found {len(services)} log service(s)")

        for svc in services:
            svc_uri = svc.get("@odata.id", "")

            # ── Fetch entries (unfiltered, capped at max-entries) ────────
            filt = LogFilter(top=args.max_entries)
            entries_resp = await logs.get_entries_async(svc_uri, filter=filt)
            if not entries_resp.success:
                print(f"    ✗ Entries HTTP {entries_resp.status_code}")
                continue

            entries = entries_resp.body.get("Members", [])
            total = entries_resp.body.get("Members@odata.count", len(entries))
            print(f"    Total entries : {total}  (showing up to {args.max_entries})")

            for entry in entries[:args.max_entries]:
                eid = entry.get("Id", "?")
                severity = entry.get("Severity", entry.get("MessageSeverity", "?"))
                msg = entry.get("Message", "")[:80]
                created = entry.get("Created", "")
                print(f"    [{eid:>6}] {severity:<12} {created[:19]}  {msg}")

            # ── Fetch single entry in detail ─────────────────────────────
            if entries:
                entry_uri = entries[0].get("@odata.id")
                if entry_uri:
                    single_resp = await logs.get_entry_async(entry_uri)
                    if single_resp.success:
                        body = single_resp.body
                        print(f"\n    Entry detail ({entry_uri}):")
                        print(f"      MessageId   : {body.get('MessageId')}")
                        print(f"      MessageArgs : {body.get('MessageArgs')}")
                        print(f"      EntryCode   : {body.get('EntryCode')}")
                        print(f"      SensorType  : {body.get('SensorType')}")

            # ── Filter by severity ───────────────────────────────────────
            print(f"\n    Fetching Critical entries …")
            crit_filter = LogFilter(severity="Critical", top=3)
            crit_resp = await logs.get_entries_async(svc_uri, filter=crit_filter)
            if crit_resp.success:
                crit_entries = crit_resp.body.get("Members", [])
                print(f"    Critical count : {len(crit_entries)}")

        # ── Clear first log (simulator may not support this) ─────────────
        if services:
            first_log_uri = services[0].get("@odata.id", "")
            print(f"\nAttempting to clear log: {first_log_uri}")
            clear_resp = await logs.clear_log_async(first_log_uri)
            if clear_resp.success:
                print("  ✓ Log cleared")
            else:
                print(f"  ✗ HTTP {clear_resp.status_code} (may not be supported by simulator)")

        print("\n✓ Log service sample complete")


if __name__ == "__main__":
    asyncio.run(main())

File: cpp/samples/08_log_service.cpp
// SPDX-License-Identifier: MIT
// Copyright (c) Microsoft Corporation. All rights reserved.

/**
 * Sample 08 — Log Service: $top / $skip / $filter and nextLink pagination
 *
 * Demonstrates:
 *   - logs.list_services()
 *   - LogQuery with top, skip, severity, message_id
 *   - logs.list_entries() — single page
 *   - logs.iter_entries()  — follow Members@odata.nextLink
 *   - logs.clear_log()
 *
 * Usage:
 *   ./08_log_service [host] [port] [max-entries]
 */
#include "redfish_sdk/client.hpp"
#include "redfish_sdk/errors.hpp"
#include <iomanip>
#include <iostream>

static void print_entry(const nlohmann::json& e) {
    std::cout << "      [" << std::left << std::setw(10) << e.value("Id","?") << "] "
              << std::setw(12) << e.value("Severity","?")
              << " " << e.value("Created","")
              << "  " << e.value("Message","").substr(0, 72) << "\n";
}

int main(int argc, char* argv[]) {
    std::string host        = (argc > 1) ? argv[1] : "127.0.0.1";
    int         port        = (argc > 2) ? std::stoi(argv[2]) : 8000;
    int         max_entries = (argc > 3) ? std::stoi(argv[3]) : 5;

    redfish::Credentials creds{"admin", "admin"};
    redfish::ConnectionConfig config;
    config.verify_tls = false;  // disable cert verification for simulator (self-signed)
    for (int i = 1; i < argc; ++i)
        if (std::string(argv[i]) == "--no-tls") config.use_tls = false;  // plain HTTP

    try {
        auto ctx  = redfish::connect(host, port, creds,
                                     redfish::AuthMode::SESSION, config);
        auto& logs = ctx->logs();

        // ── List services ─────────────────────────────────────────────
        auto svc = logs.list_services();
        if (!svc.success || !svc.body.is_object()) {
            std::cout << "No log services found\n";
            return 0;
        }
        auto members = svc.body.value("Members", nlohmann::json::array());
        std::cout << "Log services: " << members.size() << "\n";

        for (const auto& m : members) {
            std::string uri = m.value("@odata.id", "");
            std::cout << "\n  LogService: " << uri << "\n";

            // ── $top=N ───────────────────────────────────────────────
            {
                redfish::LogQuery q;
                q.top = max_entries;
                auto r = logs.list_entries(uri, q);
                auto count = r.success
                    ? r.body.value("Members@odata.count", 0) : 0;
                std::cout << "    $top=" << max_entries
                          << " : returned " << (r.success ? r.body.value("Members", nlohmann::json::array()).size() : 0)
                          << " of " << count << " total\n";
                if (r.success)
                    for (const auto& e : r.body.value("Members", nlohmann::json::array()))
                        print_entry(e);
            }

            // ── $skip + $top ─────────────────────────────────────────
            {
                redfish::LogQuery q;
                q.skip = max_entries;
                q.top  = 2;
                auto r = logs.list_entries(uri, q);
                if (r.success) {
                    auto skipped = r.body.value("Members", nlohmann::json::array());
                    std::cout << "    $skip=" << max_entries << "&$top=2 : "
                              << skipped.size() << " entry(ies)\n";
                    for (const auto& e : skipped) print_entry(e);
                }
            }

            // ── $filter=Severity ─────────────────────────────────────
            {
                redfish::LogQuery q;
                q.severity = "Warning";
                q.top      = 3;
                auto r = logs.list_entries(uri, q);
                if (r.success) {
                    auto w = r.body.value("Members", nlohmann::json::array());
                    std::cout << "    $filter=Severity eq 'Warning': "
                              << w.size() << " entry(ies)\n";
                    for (const auto& e : w) print_entry(e);
                }
            }

            // ── iter_entries: follow nextLink ─────────────────────────
            {
                std::cout << "    iter_entries (page_size=2, max 3 pages) …\n";
                int page = 0, total = 0;
                redfish::LogQuery q;
                q.top = 2;
                logs.iter_entries(uri,
                    [&](const redfish::RedfishResponse& resp) -> bool {
                        ++page;
                        int n = resp.success
                            ? static_cast<int>(resp.body.value("Members",
                                nlohmann::json::array()).size()) : 0;
                        total += n;
                        bool has_next = resp.success &&
                            resp.body.contains("Members@odata.nextLink");
                        std::cout << "      Page " << page
                                  << ": " << n << " entries"
                                  << "  nextLink=" << (has_next ? "yes" : "no")
                                  << "\n";
                        return page < 3;  // stop after 3 pages
                    }, q);
                std::cout << "      Total from " << page << " page(s): "
                          << total << " entries\n";
            }
        }

        // ── Clear first log ───────────────────────────────────────────
        if (!members.empty()) {
            std::string first_uri = members[0].value("@odata.id", "");
            std::cout << "\nClearing: " << first_uri << "\n";
            auto cr = logs.clear_log(first_uri);
            std::cout << (cr.success ? "  ✓ cleared" : "  ✗ not supported")
                      << " (HTTP " << cr.status_code << ")\n";
        }

        std::cout << "\n✓ Log service sample complete\n";
        return 0;
    } catch (const redfish::RedfishError& e) {
        std::cerr << "[ERROR] " << e.what() << "\n";
    }
    return 1;
}
